tty: Support unbuffered operation to extend usecase coverage

pfalcon opened the pull request:
The whole "tty" concept is conceived around efficient
interrupt-driven operation. However, it's beneficial to add
non interupt-driven operation under the same API:

1. Wider usecase coverage in general.
2. Allows to use the same familiar API (based on POSIX concepts)
even for UART implementations without interrupt support.
3. Allows to switch operation dynamically based on the needs.
For example, if the system is in degraded mode and interrupt
handling cannot be trusted/disabled, allows to still output
diagnostic information to user. This was the original motivation
to provide such a mode, to support logging subsystem's "panic"
mode.

To implement this feature, tty_set_rx_buf() and tty_set_tx_buf()
functions are provided, allowing to reconfigure buffers used
dynamically. If configured buffer length is 0, the operation
switched to unbuffered.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
tty: Support unbuffered operation to extend usecase coverage

The whole "tty" concept is conceived around efficient
interrupt-driven operation. However, it's beneficial to add
non interupt-driven operation under the same API:

1. Wider usecase coverage in general.
2. Allows to use the same familiar API (based on POSIX concepts)
even for UART implementations without interrupt support.
3. Allows to switch operation dynamically based on the needs.
For example, if the system is in degraded mode and interrupt
handling cannot be trusted/disabled, allows to still output
diagnostic information to user. This was the original motivation
to provide such a mode, to support logging subsystem's "panic"
mode.

To implement this feature, tty_set_rx_buf() and tty_set_tx_buf()
functions are provided, allowing to reconfigure buffers used
dynamically. If configured buffer length is 0, the operation
switched to unbuffered.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

jarz-nordic commented on subsys/console/tty.c:
```diff
@@ -149,12 +160,54 @@ static int tty_getchar(struct tty_serial *tty)
+			 * directly if none.
+			 */
+			if (out_size == 0) {
+				errno = res;
```
Two questions here:
1. where `errno` is defined?
2. Do you assume that tty module will not be multiinstance?

pfalcon replied:
@jarz-nordic, this is hard(ish) one.

1. First of all, it follows the pattern used by other code in this module.
2. And the code in this module does that, because the requirement is to be POSIX compatible.
3. POSIX defines `errno`, in `<errno.h>` (we define it "somewhere").
4. `errno` is a thread-local variable.

So, it wasn't that hard after all. The only realistic way to perform *concurrent* operations (and that's your concern) on multiple instances is via different threads, and POSIX has that covered. Otherwise, this API is not intended to be called from ISRs (but if one does, it might even not crash, just some aspects would be funky, like competing for errno writes).

jarz-nordic replied:
@pfalcon : Thank you for explanation.

pfalcon committed:
tty: Remove buffer params from tty_init()

Let's have more orthogonal and cleaner API, where buffers are
configured by tty_set_rx_buf/tty_set_tx_buf, and only them. It
means that newly initialized tty starts in unbuffered mode, which
is somewhat a sidestep from a main usecase behind tty, which is
buffered operation, but again, having a cleaner API (and good
docs, explaining users how it should be and what they should do)
prevails.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
> void tty_init(struct tty_serial *tty, struct device *uart_dev);

Now done.

Also quickly tested on qemu_x86 that console/echo didn't regress, but needs more thorough testing of course.

pfalcon committed:
console: Kconfig: Update for recent changes

Buffer sizes aren't required to be power of 2 for a while. Describe
that by setting buffers sizes to 0, one can get non interrupt driven
operation.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Ok, made a couple of tweaks and tested on real hardware. Should be ready to merge now.

Specifically, tested that with
~~~
CONFIG_CONSOLE_GETCHAR_BUFSIZE=0
CONFIG_CONSOLE_PUTCHAR_BUFSIZE=0
~~~

(i.e. disabling buffering and interrupt-driven operation)

Any clipboard paste into the UART connection beyond a handlful of chars leads to char drops - which was the original reason I embarked to write these console routines in the first place 2.5 years ago.

@nordic-krch, @jarz-nordic, @mike-scott, et al: Please review.

jarz-nordic commented on subsys/console/tty.c:
```diff
@@ -149,12 +160,59 @@ static int tty_getchar(struct tty_serial *tty)
+			size--;
+		}
+
+		if (size == 0 || (timeout != K_FOREVER && timeout-- == 0)) {
```
This conditions is strange. It is obvious that if `timeout == 0` it is `!= K_FOREVER`.
Didn't you mean:
`if (size == 0 || timeout == K_FOREVER || timeout-- == 0)`

pfalcon replied:
Well, quite classy C condition ;-). The purpose of `timeout != K_FOREVER &&` is to short-circuit evaluation of `timeout-- == 0` which has a side effect in it.

pfalcon commented:
@jarz-nordic: Thanks for review!

@nordic-krch, @carlescufi : Ping.

nordic-krch commented:
@pfalcon sorry for the delay. I forgot to press 'finish review' long time ago and it was pending in the browser.

pfalcon commented on subsys/console/tty.c:
```diff
@@ -149,12 +160,54 @@ static int tty_getchar(struct tty_serial *tty)
+	while (size) {
+		u8_t c;
+		res = uart_poll_in(tty->uart_dev, &c);
+		if (res <= -2) {
```
> why -2 here?

Based on description of uart_poll_in() https://github.com/zephyrproject-rtos/zephyr/blob/master/include/uart.h#L280 . I.e. -1 is transient condition aka EAGAIN. Anything below is real error.

It probably would catch less attention if was written as `if (res < -1)`, but disguising to not catch attention is not the aim here ;-).

pfalcon commented on subsys/console/tty.c:
```diff
@@ -149,12 +160,54 @@ static int tty_getchar(struct tty_serial *tty)
+			break;
+		}
+
+		k_sleep(1);
```
@nordic-krch :

> this triggers rescheduling. If blocking tty is used in panic then that should not happen. Especially that call will most likely happen from irq (e.g. hardfault) and k_sleep is for thread only (asserts in irq). k_busy_wait would be better option.

Did you notice that this k_sleep() is in read() function? You're not interested/not going to call read() for logging, right? Otherwise the idea of k_sleep() is exactly to avoid burning CPU at 100% usage while waiting for UART input in the not interrupt driven case, letting it sleep or process other tasks for a bit.

pfalcon commented on subsys/console/tty.c:
```diff
@@ -104,6 +104,17 @@ ssize_t tty_write(struct tty_serial *tty, const void *buf, size_t size)
+		out_size = size;
+
+		while (size--) {
+			uart_poll_out(tty->uart_dev, *p++);
```
> i wonder what we should do if flow control is on and receiver is blocking transmission

Well, yeah, flow control support in both drivers and APIs is sparse at best. I indeed to just what driver allows me to, I even describe that non-buffered write operations are implicitly blocking and timeout setting don't apply to them.

So yes, once flow control support is better implemented/validated in drivers, we would need to fix it on their level (new method? reuse timeout set by the new UART API? dunno).

pfalcon commented:
@nordic-krch 
> sorry for the delay. I forgot to press 'finish review' long time ago and it was pending in the browser.

np. I also missed some comments, as some of them now apply to the changed code. I believe I responded to all now, let me know if you suggest some changes.
